aws/credentials/processcreds/provider.go

```diff
@@ -0,0 +1,476 @@
+	Version         int
+	AccessKeyID     string `json:"AccessKeyId"`
+	SecretAccessKey string
+	SessionToken    string
+	Expiration      *time.Time
+}
+
+// Retrieve executes the 'credential_process' and returns the credentials.
+func (p *ProcessProvider) Retrieve() (credentials.Value, error) {
+	out, err := p.executeCredentialProcess()
+	if err != nil {
+		return credentials.Value{ProviderName: ProviderName}, err
+	}
+
+	// Serialize and validate response
+	resp := &credentialProcessResponse{}
+	if err = json.Unmarshal(out, resp); err != nil {
+		return credentials.Value{ProviderName: ProviderName}, awserr.New(
+			ErrCodeProcessProviderParse,
+			fmt.Sprintf("%s: %s", ErrMsgProcessProviderParse, string(out)),
+			err)
+	}
+
+	if resp.Version != 1 {
+		return credentials.Value{ProviderName: ProviderName}, awserr.New(
+			ErrCodeProcessProviderVersion,
+			ErrMsgProcessProviderVersion,
+			nil)
+	}
+
+	if len(resp.AccessKeyID) == 0 {
+		return credentials.Value{ProviderName: ProviderName}, awserr.New(
+			ErrCodeProcessProviderRequired,
+			ErrMsgProcessProviderMissKey,
+			nil)
+	}
+
+	if len(resp.SecretAccessKey) == 0 {
+		return credentials.Value{ProviderName: ProviderName}, awserr.New(
+			ErrCodeProcessProviderRequired,
+			ErrMsgProcessProviderMissSecret,
+			nil)
+	}
+
+	// Handle expiration
+	p.staticCreds = resp.Expiration == nil
+	if resp.Expiration != nil {
+		p.SetExpiration(*resp.Expiration, p.ExpiryWindow)
+	}
+
+	return credentials.Value{
+		ProviderName:    ProviderName,
+		AccessKeyID:     resp.AccessKeyID,
+		SecretAccessKey: resp.SecretAccessKey,
+		SessionToken:    resp.SessionToken,
+	}, nil
+}
+
+// IsExpired returns true if the credentials retrieved are expired, or not yet
+// retrieved.
+func (p *ProcessProvider) IsExpired() bool {
+	if p.staticCreds {
+		return false
+	}
+	return p.Expiry.IsExpired()
+}
+
+func findInSlice(needle string, haystack []string) bool {
+	for _, straw := range haystack {
+		if strings.HasPrefix(straw, needle) {
+			return true
+		}
+	}
+	return false
+}
+
+func getProcessProviderEnv(existEnv []string) []string {
+	if !findInSlice(AlreadyPreparedCommandKey, existEnv) {
+		existEnv = append(existEnv, fmt.Sprintf(
+			"%s=%s",
+			AlreadyPreparedCommandKey,
+			"true"))
+
+		if path, ok := os.LookupEnv("PATH"); !ok || len(path) < 1 {
+			if runtime.GOOS == GoOSWindowsKey {
+				existEnv = append(existEnv, fmt.Sprintf(
+					"%s=%s",
+					"PATH",
+					"C:\\Windows\\system32"))
+			} else {
+				existEnv = append(existEnv, fmt.Sprintf("%s=%s", "PATH", "/bin"))
+			}
+		}
+
+		existEnv = append(os.Environ(), existEnv...)
+	}
+
+	return existEnv
+}
+
+func getProcessProviderArgs(command string) []string {
+	var cmdArgs []string
+	if runtime.GOOS == GoOSWindowsKey {
+		if comspec, ok := os.LookupEnv("ComSpec"); ok {
+			cmdArgs = []string{comspec, "/C"}
+		} else {
+			cmdArgs = []string{"C:\\Windows\\system32\\cmd.exe", "/C"}
+		}
+	} else {
+		cmdArgs = []string{"/bin/sh", "-c"}
```

Comment:
@YakDriver  could you comment on why the OS's shell needs to be added to the command arguments, as apposed to just executing the command directly with `exec.Command`? If there are specific shell environment needed I think the user would be responsible for providing it.

Reply:
This avoids the problem of splitting into command and arguments. Splitting on spaces is prone to problems with quotes and quickly turns into a complicated problem. Using a subshell like this is pretty common solution to the problem:

1. [Unexpected behavior of exec.Command argument parsing](https://groups.google.com/d/msg/golang-nuts/qYVlLVR5504/o-Td7DU6qsIJ)
1. [Splitting up quoted/escaped command line arguments](https://www.perlmonks.org/?node_id=1074460)
1. [Core code of Terraform's local exec provisioner](https://github.com/hashicorp/terraform/blob/d4ac68423c4998279f33404db46809d27a5c2362/builtin/provisioners/local-exec/resource_provisioner.go#L79-L83)